ca/ca.go

```diff
@@ -427,13 +432,61 @@ func (ca *CertificateAuthorityImpl) IssuePrecertificate(ctx context.Context, iss
 		return nil, err
 	}
 
-	precertDER, err := ca.issuePrecertificateInner(ctx, issueReq, serialBigInt, validity, precertType)
-	if err != nil {
-		return nil, err
+	if features.Enabled(features.PrecertificateOCSP) {
+		serialHex := core.SerialToString(serialBigInt)
+		nowNanos := ca.clk.Now().UnixNano()
+		expiresNanos := validity.NotAfter.UnixNano()
+		_, err = ca.sa.AddSerial(ctx, &sapb.AddSerialRequest{
+			Serial:  &serialHex,
+			RegID:   issueReq.RegistrationID,
+			Created: &nowNanos,
+			Expires: &expiresNanos,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		precertDER, err := ca.issuePrecertificateInner(ctx, issueReq, serialBigInt, validity, precertType)
+		if err != nil {
+			return nil, err
+		}
+
+		ocspResp, err := ca.GenerateOCSP(ctx, core.OCSPSigningRequest{
+			CertDER: precertDER,
+			Status:  "good",
+		})
+		if err != nil {
+			err = berrors.InternalServerError(err.Error())
+			ca.log.AuditInfof("OCSP Signing failure: serial=[%s] err=[%s]", core.SerialToString(serialBigInt), err)
+		}
```

Comment:
Is there a missing return here, or are we not returning because we've already issued the precertificate, and as such are deciding to forge ahead regardless? I think if this is intentional we'd still never end up with a ocsp response as the `AddCertificate` call in `generateOCSPAndStoreCertificate` would fail to add the new row containing the OCSP response generated at that step.

Reply:
Also seems like `generateOCSPAndStoreCertificate` should be modified to not attempt to generate the OCSP if the flag is enabled, currently this will cause us to do an extra signature per-certificate we issue.

Reply:
(might also make sense to update the AddCertificate RPC with a flag that tells the SA to just not bother attempting to add the certificateStatus row to save unnecessary duplicate writes.)

Reply:
This is a rare case (zero instances in the logs), but I think what we want to do here is the same as what we do in the final certificate case: write the cert to the DB anyhow (since we signed it already), and let the ocsp-updater fix up the OCSP status later. Filed an issue to fix ocsp-updater so it correctly finds the status: https://github.com/letsencrypt/boulder/issues/4426.

Reply:
> Also seems like generateOCSPAndStoreCertificate should be modified to not attempt to generate the OCSP if the flag is enabled, currently this will cause us to do an extra signature per-certificate we issue.
> (might also make sense to update the AddCertificate RPC with a flag that tells the SA to just not bother attempting to add the certificateStatus row to save unnecessary duplicate writes.)

We do about 30x as many signatures for OCSP as we do for issuances, so doing one extra signature at issuance time is not a big deal IMO, and will only be the case for a week or so after deployment, till we go through and clean up the old code. Same with the flag to avoid the write to the certificateStatus row, although there I suppose there's some risk that we could have lower availability because the extra query nudges us slightly outside our timeout.

Reply:
I'm not really convinced by this argument. Leaving code around that has duplicated unnecessary functionality because we'll probably get around to removing it soon seems like a pretty bad anti-pattern. Past experience has shown that quite often we're not able to remove code in a single deploy cycle for numerous reasons. The downside here of adding a feature check around the block in question seems almost nil, there is some minor churn in terms of adding the check but if anything that'll make removing the code, and the review of that removal, even simpler.